File: rust/autd3-driver/src/fpga/defined.rs
/*
 * File: fpga_defined.rs
 * Project: src
 * Created Date: 02/05/2022
 * Author: Shun Suzuki
 * -----
 * Last Modified: 03/03/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2022 Shun Suzuki. All rights reserved.
 *
 */

use std::f64::consts::PI;

pub const FPGA_CLK_FREQ: usize = 163840000;

pub const FOCUS_STM_FIXED_NUM_UNIT: f64 = 0.025; //mm

#[derive(Clone, Copy, Debug)]
pub struct Amp {
    value: f64,
}

impl Amp {
    pub fn new(value: f64) -> Self {
        Self {
            value: value.clamp(0., 1.),
        }
    }

    pub fn value(&self) -> f64 {
        self.value
    }
}

#[derive(Clone, Copy, Debug)]
pub struct Phase {
    value: f64,
}

impl Phase {
    pub fn new(value: f64) -> Self {
        Self { value }
    }

    pub fn value(&self) -> f64 {
        self.value
    }
}

#[derive(Clone, Copy, Debug)]
pub struct Drive {
    pub phase: Phase,
    pub amp: Amp,
}

#[derive(Clone, Copy, Debug)]
#[repr(C)]
pub struct LegacyDrive {
    pub phase: u8,
    pub duty: u8,
}

impl LegacyDrive {
    pub fn to_phase(d: &Drive) -> u8 {
        (((d.phase.value() / (2.0 * PI) * 256.0).round() as i32) & 0xFF) as _
    }

    pub fn to_duty(d: &Drive) -> u8 {
        (510.0 * d.amp.value().clamp(0., 1.).asin() / PI).round() as _
    }

    pub fn set(&mut self, d: &Drive) {
        self.duty = Self::to_duty(d);
        self.phase = Self::to_phase(d);
    }
}

#[derive(Clone, Copy, Debug)]
#[repr(C)]
pub struct AdvancedDrivePhase {
    pub phase: u16,
}

impl AdvancedDrivePhase {
    pub fn to_phase(d: &Drive, cycle: u16) -> u16 {
        ((d.phase.value() / (2.0 * PI) * cycle as f64).round() as i32).rem_euclid(cycle as i32) as _
    }

    pub fn set(&mut self, d: &Drive, cycle: u16) {
        self.phase = Self::to_phase(d, cycle);
    }
}

#[derive(Clone, Copy, Debug)]
#[repr(C)]
pub struct AdvancedDriveDuty {
    pub duty: u16,
}

impl AdvancedDriveDuty {
    pub fn to_duty(d: &Drive, cycle: u16) -> u16 {
        (cycle as f64 * d.amp.value().clamp(0., 1.).asin() / PI).round() as _
    }

    pub fn set(&mut self, d: &Drive, cycle: u16) {
        self.duty = Self::to_duty(d, cycle);
    }
}

#[derive(Default)]
#[repr(C)]
pub struct FPGAInfo {
    info: u8,
}

impl FPGAInfo {
    pub fn is_thermal_assert(&self) -> bool {
        (self.info & 0x01) != 0
    }
}

File: tests/driver/common/fpga/defined.cpp
// File: defined.cpp
// Project: fpga
// Created Date: 30/11/2022
// Author: Shun Suzuki
// -----
// Last Modified: 20/03/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2022 Shun Suzuki. All rights reserved.
//

#if _MSC_VER
#pragma warning(push)
#pragma warning(disable : 26439 26495 26812)
#endif
#include <gtest/gtest.h>
#if _MSC_VER
#pragma warning(pop)
#endif

#include "autd3/driver/fpga/defined.hpp"

using autd3::driver::Drive;
using autd3::driver::pi;

TEST(DriverCommonFPGADefined, LegacyDrive) {
  using autd3::driver::LegacyDrive;

  ASSERT_EQ(sizeof(LegacyDrive), 2);

  Drive s{0, 0};
  uint8_t d[2]{};

  reinterpret_cast<LegacyDrive*>(&d)->set(s);
  ASSERT_EQ(d[0], 0);
  ASSERT_EQ(d[1], 0);

  s.phase = pi;
  s.amp = 0.5;
  reinterpret_cast<LegacyDrive*>(&d)->set(s);
  ASSERT_EQ(d[0], 128);
  ASSERT_EQ(d[1], 85);

  s.phase = 2.0 * pi;
  s.amp = 1.0;
  reinterpret_cast<LegacyDrive*>(&d)->set(s);
  ASSERT_EQ(d[0], 0);
  ASSERT_EQ(d[1], 255);

  s.phase = 3.0 * pi;
  s.amp = 1.5;
  reinterpret_cast<LegacyDrive*>(&d)->set(s);
  ASSERT_EQ(d[0], 128);
  ASSERT_EQ(d[1], 255);

  s.phase = -pi;
  s.amp = -1;
  reinterpret_cast<LegacyDrive*>(&d)->set(s);
  ASSERT_EQ(d[0], 128);
  ASSERT_EQ(d[1], 0);
}

TEST(DriverCommonFPGADefined, Phase) {
  using autd3::driver::AdvancedDrivePhase;

  ASSERT_EQ(sizeof(AdvancedDrivePhase), 2);

  Drive s{0, 0};
  uint16_t d{0};

  {
    constexpr uint16_t cycle = 4096;
    s.phase = 0.0;
    reinterpret_cast<AdvancedDrivePhase*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 0);

    s.phase = pi;
    reinterpret_cast<AdvancedDrivePhase*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 2048);

    s.phase = 2.0 * pi;
    reinterpret_cast<AdvancedDrivePhase*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 0);

    s.phase = 3.0 * pi;
    reinterpret_cast<AdvancedDrivePhase*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 2048);

    s.phase = -pi;
    reinterpret_cast<AdvancedDrivePhase*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 2048);
  }

  {
    constexpr uint16_t cycle = 2000;
    s.phase = 0.0;
    reinterpret_cast<AdvancedDrivePhase*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 0);

    s.phase = pi;
    reinterpret_cast<AdvancedDrivePhase*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 1000);

    s.phase = 2.0 * pi;
    reinterpret_cast<AdvancedDrivePhase*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 0);

    s.phase = 3.0 * pi;
    reinterpret_cast<AdvancedDrivePhase*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 1000);

    s.phase = -pi;
    reinterpret_cast<AdvancedDrivePhase*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 1000);
  }
}

TEST(DriverCommonFPGADefined, Duty) {
  using autd3::driver::AdvancedDriveDuty;

  ASSERT_EQ(sizeof(AdvancedDriveDuty), 2);

  Drive s{0, 0};
  uint16_t d{0};

  {
    constexpr uint16_t cycle = 4096;

    s.amp = 0.0;
    reinterpret_cast<AdvancedDriveDuty*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 0);

    s.amp = 0.5;
    reinterpret_cast<AdvancedDriveDuty*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 683);

    s.amp = 1.0;
    reinterpret_cast<AdvancedDriveDuty*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 2048);

    s.amp = 1.5;
    reinterpret_cast<AdvancedDriveDuty*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 2048);

    s.amp = -1;
    reinterpret_cast<AdvancedDriveDuty*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 0);
  }

  {
    constexpr uint16_t cycle = 2000;

    s.amp = 0.0;
    reinterpret_cast<AdvancedDriveDuty*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 0);

    s.amp = 0.5;
    reinterpret_cast<AdvancedDriveDuty*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 333);

    s.amp = 1.0;
    reinterpret_cast<AdvancedDriveDuty*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 1000);

    s.amp = 1.5;
    reinterpret_cast<AdvancedDriveDuty*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 1000);

    s.amp = -1;
    reinterpret_cast<AdvancedDriveDuty*>(&d)->set(s, cycle);
    ASSERT_EQ(d, 0);
  }
}

TEST(DriverCommonFPGADefined, FPGAInfo) {
  using autd3::driver::FPGAInfo;

  ASSERT_EQ(sizeof(FPGAInfo), 1);

  FPGAInfo info(0);
  ASSERT_FALSE(info.is_thermal_assert());

  info = FPGAInfo(1);
  ASSERT_TRUE(info.is_thermal_assert());

  info = FPGAInfo(2);
  ASSERT_FALSE(info.is_thermal_assert());
}
